src/pybind/cephfs/cephfs.pyx

```diff
@@ -2878,6 +2878,297 @@ cdef class LibCephFS(object):
                          f'file at path {trash_path}: {e}')
                 raise
 
+    def cptree(self, src_path, dst_path, should_cancel, suppress_errors=False):
+        '''
+        Copy entire file hierarchy under src using depth-first (to prevent
+        excessive memory consumption) and non-recursive (to prevent hitting
+        Python's max recursion limit error) approach
+
+        If src is regfile, symlink or something else, copy it to dst and return.
+        '''
+        if isinstance(src_path, str):
+            src_path = src_path.encode('utf-8')
+        if isinstance(dst_path, str):
+            dst_path = dst_path.encode('utf-8')
+
+        # stx_b = statx buffer
+        stx_b = self.statx(src_path, CEPH_STATX_MODE, AT_SYMLINK_NOFOLLOW)
+        if stat.S_ISDIR(stx_b['mode']):
+            NonRecursiveCptree(self, src_path, dst_path, should_cancel,
+                               suppress_errors).cptree()
+        elif stat.S_ISREG(stx_b['mode']):
+            mode = stx_b['mode'] & ~stat.S_IFMT(stx_b['mode'])
+
+            src_file_name = os.path.basename(src_path)
+            dst_path = os.path.join(dst_path, src_file_name)
+
+            copy_reg_file(self, src_path, dst_path, mode)
+        elif stat.S_ISLNK(stx_b['mode']):
+            mode = stx_b['mode'] & ~stat.S_IFMT(stx_b['mode'])
+
+            src_slink_name = os.path.basename(src_path)
+            dst_path = os.path.join(dst_path, src_slink_name)
+
+            copy_sym_link(self, src_path, dst_path, mode)
```

Comment:
I don't understand special casing this here. The `CopyTreeWorker` should just handle that.

When you finally use `openat` style API, you can change your constructor to:

```
    def __init__(self, fs, srcdir: int, path1: str, destdir: int, dst_path2: str, should_cancel, suppress_errors=False):
```